py3-h5py: Multi-versionize py3-h5py and deps

justinvreeland opened the pull request:
Build py3-h5py and deps for multiple versions of python3

justinvreeland commented on py3-cython.yaml:
```diff
@@ -40,17 +40,39 @@ subpackages:
+        - py${{range.key}}-${{vars.pypi-package}}-bin
+      provides:
+        - py3-${{vars.pypi-package}}
+        - cython
```
Tagged this on the other PR but this shouldn't provide the models just the binaries.

pnasrat replied:
`py3-build.yaml` has the provides here to be `py3-${{vars.pypi-package}}` for the bin subpackage. I don't know if we might want this to be

`cython=${{package.full-version}}" as we have a `cython-0` version stream also

justinvreeland replied:
I think we want `cython-bin` subpackages to be dependent on the corresponding non-bin subpackages.   And we want the `-bin` packages to have the provides for `cython` that way installing the modules doesn't necessarily install the binaries but installing a specific binary version gets you the corresponding modules.

I don't think we want to provide a new stream though.  No-one should be targeting a specific stream other than the 0 stream and it's not being used by anything anymore. The package it was added for doesn't have problems building with newer versions of cython.

smoser replied:
How would a multi-versioned py3-foo.yaml use cython?

As I understand it, something using cython is going to need a py3.XX-cython and a /usr/bin/cython.
I ran into this just recently in py3-maturin also.

How about this:
1. make py3-cython package build a non-versioned '/usr/bin/cython' with the package name 'cython'
2. remove the cython-0 package
3. make py3.XX-cython depend on 'cython'

Then, to use cython in a build the package just needs to add `py3-cython` or `py3-supported-cython`.

justinvreeland replied:
> How would a multi-versioned py3-foo.yaml use cython?

The same way they do now by having a dep on cython? This only changes if their package imports cython from the building interpreter not if they use a compiler. 

> remove the cython-0 package

I missed 2 packages that are using cython~0 so getting rid of it might not be possible right now. I'll check if they can still need that version.

> Then, to use cython in a build the package just needs to add py3-cython or py3-supported-cython

We already have people depending more people depending on `cython` than `py3-cython` and I don't see an advantage to having a real `cython` package rather than bin packages that provide `cython` and letting apk sort out which one to use. As I understand it the dependency solver will choose based on package priority so it's already going to end up with the newest interpreter. I do see potential advantage to being able to create an environment with only a single python interpreter and all the expected pieces available.

justinvreeland commented on py3-meson-python.yaml:
```diff
@@ -10,10 +10,21 @@ package:
       - py3-colorama
       - py3-packaging
-      - py3-pyproject-metadata
+      - py3-pyproject-metadat
```
spelling error? ahh need to clean up the runtime deps.

justinvreeland reviewed (commented):
Draft of py3-p5py for context for python and dumpy changes. Still has some work updating the various blocks to match best practices.

pnasrat commented on py3-cython.yaml:
```diff
@@ -40,17 +40,38 @@ subpackages:
+    pipeline:
+      - runs: |
+          mkdir -p ${{targets.contextdir}}/usr
+          mv ${{targets.contextdir}}/../py${{range.key}}-${{vars.pypi-package}}/usr/bin ${{targets.contextdir}}/usr
```
Add a -bin subpackage inline test perhaps just `cython --version`

justinvreeland replied:
I'm going to implement these changes over on https://github.com/wolfi-dev/os/pull/26844 because h5py still has a fair amount of deps to be fixed up and yarl can reasonably land soon. 

Alternatively I can split off cython and numpy and pickup pandas as well and do those in one PR which tackles some similar packages on our list.

Commit:
py3-more-itertools: Multi-versionize

- Update the build
- Add tests

Commit:
py3-jaraco.functools: New package!

Added as a dep

Commit:
py3-jaraco.context: New package!

Added as a dep

Commit:
py3-meson-python: Multi-versionize

- Update build
- Add supported packages for use in builds
- Add tests

Commit:
py3-poetry-core: Multi-versionize!

- Build multiple versions
- Build supported packages to use while building
- Add tests
- Remove generated comment since it's significantly changed

Commit:
py3-ptyprocess: Multi-versionize ptyprocess

- Update Build
- Add tests
- Add supported metapackage
- Remove generation comment since it's significantly changed

Commit:
py3-cffi: Multi versionize

- Update build
- Add tests
- Remove generation comment
- Add metapackage for builds

Commit:
New pipeline for build and installer

Also a usage!

Commit:
py3-tomlkit: Multi-versionize tomlkit

- Update build
- remove genrated comment
- Add tests

Commit:
py3-xattr: Mult-versionize

- Update build
- Add tests
- Remove generation comment
- Add metapackage for builds

Commit:
py3-requests: Multi-versionize

- Update build
- Add tests
- Remove generation comment
- Add metapackage for builds

Commit:
py3-xattr fixup

Commit:
py3-shellingham: Multi-versionize

- Update build
- Add tests
- Remove generation comment

Commit:
py3-platformdirs: Multi-versionize

- Update build
- Add tests
- Remove generation comment

Commit:
fixup shellingham

Commit:
py3-msgpack: Multi-versionize

- Update build
- remove genrated comment
- Add tests

Commit:
py3-maturin misc updates

Commit:
py3-charset-normalizer: Multi-versionize

Commit:
py3-requests: Multi-versionize

smoser commented on pipelines/py/build-installer.yaml:
```diff
@@ -0,0 +1,52 @@
+      echo "$py is $pyver with site_packages dir '$sitepkgd'"
+      vr $py -m build --wheel --verbose --no-isolation
+      vr $py -m installer -d $root dist/*.whl
+      vr find $root -name "*.pyc" -exec rm -rf '{}' +
```
this is wrong. we do not want to remove pyc files.

justinvreeland replied:
Ahh OK I can remove that, It does seem to be fairly common in our files though

Commit:
py3-cachecontrol: Multi version updates

Commit:
py3-maturin: Updates

Commit:
py3-virtualenv: Multi-versionize

smoser commented on py3-distlib.yaml:
```diff
@@ -28,31 +34,34 @@ pipeline:
+      provides:
+        - py3-${{vars.pypi-package}}
+    pipeline:
+      - uses: py/build-installer
```
took this file as it is right now, and added py3-supported-pip to the envionment.contents.packages list and then it works with 'pip-build-install'.

I suspect the same is true for many of the other ones.

Its annoying that you have to get the dependencies rigth for the pipeline to build.  That is fallout of the pipeline being "multi-version".  It _could_ say it depends on `py3-supported-build`.  maybe that is the rigth way to do things, but that would cause pain for the non-versioned path.... they'd get all the python3-base things installed for a build of just the latest python3.

justinvreeland replied:
> I suspect the same is true for many of the other ones.
Yeah, I've only encountered one that it didn't work for but encountering that one made me think I was better off using build-installer for things that already specified it and thus I made the pipeline. I'm happy to only change it when something doesn't seem to work with the typical one we use.

Commit:
py3-jaraco.{classes,context}: Multiversionize

Commit:
misc 'yam' fixes

smoser commented on py3-shellingham.yaml:
```diff
@@ -28,13 +33,31 @@ pipeline:
+      provides:
+        - py3-${{vars.pypi-package}}
+    pipeline:
+      - uses: py/build-installer
```
this is really wierd. you dropped py3-build and installer from the environment.contents.packages but then use build-installer ? how does that even work?

justinvreeland replied:
Probably a mistake, I added `build-installer` pretty late in going through these deps when once of my packages didn't install correctly with the pip pipeline and started retroactively converting the old ones back.

smoser reviewed (changes requested):
Probably best to split this up.  Its just really big at this point.  Maybe 3 or 4 at a time?

justinvreeland commented:
> Probably best to split this up. Its just really big at this point. Maybe 3 or 4 at a time?

Yeah I do intend of splitting this apart it much, much smaller pieces. But this is roughly the stack I had once I got all of the packages building and started cleaning things up.
